lib/SILOptimizer/Transforms/SILMem2Reg.cpp

```diff
@@ -401,7 +403,95 @@ static void endLexicalLifetimeBeforeInst(AllocStackInst *asi,
   SILLocation loc =
       RegularLocation::getAutoGeneratedLocation(beforeInstruction->getLoc());
   builder.createEndBorrow(loc, values.borrow);
-  builder.emitDestroyValueOperation(loc, values.stored);
+  builder.emitDestroyValueOperation(loc, values.stored->getOwnershipKind() ==
+                                                 OwnershipKind::Guaranteed
+                                             ? values.copy
+                                             : values.stored);
+}
+
+/// Returns true if \p src is a function parameter or a projection of it.
+static bool shouldExtendLifetimeForStoreBorrowSource(SILValue src) {
+  if (isa<SILFunctionArgument>(src))
+    return false;
+  if (Projection::isObjectProjection(src)) {
+    return shouldExtendLifetimeForStoreBorrowSource(
+        cast<SingleValueInstruction>(src)->getOperand(0));
+  }
+  return true;
+}
+
+/// Extends the lifetime of \p sbi source by creating a copy and a borrow. i.e:
+/// Transform store_borrow %src to %asi:
+/// %copy = copy_value %src
+/// %borrow = begin_borrow %copy
+static StorageStateTracking<LiveValues>
+extendLifetimeForStoreBorrowSource(AllocStackInst *asi, StoreBorrowInst *sbi) {
+  assert(shouldExtendLifetimeForStoreBorrowSource(sbi->getSrc()));
+
+  BeginBorrowInst *bbi = nullptr;
+  CopyValueInst *cvi = nullptr;
+  SILBuilderWithScope::insertAfter(sbi, [&](SILBuilder &builder) {
+    SILLocation loc = RegularLocation::getAutoGeneratedLocation(sbi->getLoc());
+    cvi = builder.createCopyValue(loc, sbi->getSrc());
+    bbi = builder.createBeginBorrow(loc, cvi, asi->isLexical());
+  });
+  StorageStateTracking<LiveValues> vals = {{sbi->getSrc(), bbi, cvi},
+                                           /*isStorageValid=*/true};
+  return vals;
+}
+
+/// Returns true, if the stored guaranteed value has a copy and borrow whose
+/// lifetime needs to be ended.
+static bool shouldEndExtendedLifetime(LiveValues values) {
+  assert(values.stored->getOwnershipKind() == OwnershipKind::Guaranteed);
+  return values.borrow && values.copy;
+}
+
+/// Ends lifetime of the borrow and copy of the stored value inserting an
+/// end_borrow and destroy_value before \p endBefore.
+static void endExtendedLifetime(SILInstruction *endBefore,
+                                SILBuilderContext &ctx, LiveValues values) {
+  assert(endBefore);
+  assert(shouldEndExtendedLifetime(values));
+  assert(values.stored->getOwnershipKind() == OwnershipKind::Guaranteed);
+
+  auto builder = SILBuilderWithScope(endBefore, ctx);
+  SILLocation loc =
+      RegularLocation::getAutoGeneratedLocation(endBefore->getLoc());
+  builder.createEndBorrow(loc, values.borrow);
+  builder.emitDestroyValueOperation(loc, values.copy);
```

Comment:
Similarly, we shouldn't need the `destroy_value` for a `store_borrow` location. Basically, a store-borrow location always has guaranteed ownership.
We just need to make sure that the store-borrow lifetime ends at the point where we originally had `end_borrow` instructions.